nips2018/mnist_hypersphere.py

```diff
@@ -0,0 +1,77 @@
+"""Trains a simple convnet on the MNIST dataset.
+Convolutional Layers are trained on the Hypersphere.
+
+Based on the keras mnist cnn example.
+"""
+
+from __future__ import print_function
+import keras
+from keras.datasets import mnist
+from keras.models import Sequential
+from keras.layers import Dense, Dropout, Flatten
+from keras.layers import Conv2D, MaxPooling2D
+from keras import backend as K
+from geomstats.hypersphere import Hypersphere
+
+batch_size = 128
+num_classes = 10
+epochs = 12
+
+# input image dimensions
+img_rows, img_cols = 28, 28
+
+# the data, split between train and test sets
+(x_train, y_train), (x_test, y_test) = mnist.load_data()
+
+if K.image_data_format() == 'channels_first':
+    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
+    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
+    input_shape = (1, img_rows, img_cols)
+else:
+    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
+    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
+    input_shape = (img_rows, img_cols, 1)
+
+x_train = x_train.astype('float32')
+x_test = x_test.astype('float32')
+x_train /= 255
+x_test /= 255
+print('x_train shape:', x_train.shape)
+print(x_train.shape[0], 'train samples')
+print(x_test.shape[0], 'test samples')
+
+# convert class vectors to binary class matrices
+y_train = keras.utils.to_categorical(y_train, num_classes)
+y_test = keras.utils.to_categorical(y_test, num_classes)
+
+model = Sequential()
+
+kernel_size = (3, 3)
+hypersphere_dimension = kernel_size[0] * kernel_size[1] - 1
+model.add(Conv2D(32, kernel_size=kernel_size,
+                 activation='relu',
+                 input_shape=input_shape,
+                 kernel_manifold=Hypersphere(hypersphere_dimension)))
```

Comment:
Don't you want to instantiate one hypersphere and then pass it as an parameter, instead of instantiating it twice?

`S2 = Hypersphere(hypersphere_dimension)` and then `kernel_manifold=S2` wherever you need it?